libcudacxx/include/cuda/__iterator/zip_iterator.h

```diff
@@ -0,0 +1,591 @@
+#include <cuda/std/__iterator/incrementable_traits.h>
+#include <cuda/std/__iterator/iter_move.h>
+#include <cuda/std/__iterator/iter_swap.h>
+#include <cuda/std/__iterator/iterator_traits.h>
+#include <cuda/std/__ranges/concepts.h>
+#include <cuda/std/__type_traits/common_type.h>
+#include <cuda/std/__type_traits/make_unsigned.h>
+#include <cuda/std/__type_traits/remove_cvref.h>
+#include <cuda/std/__utility/forward.h>
+#include <cuda/std/__utility/integer_sequence.h>
+#include <cuda/std/__utility/move.h>
+#include <cuda/std/__utility/pair.h>
+#include <cuda/std/tuple>
+
+#include <cuda/std/__cccl/prologue.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+template <class... _Iterators>
+struct __tuple_or_pair_impl
+{
+  using type = _CUDA_VSTD::tuple<_Iterators...>;
+};
+
+template <class _Iterator1, class _Iterator2>
+struct __tuple_or_pair_impl<_Iterator1, _Iterator2>
+{
+  using type = _CUDA_VSTD::pair<_Iterator1, _Iterator2>;
+};
+
+template <class... _Iterators>
+using __tuple_or_pair = typename __tuple_or_pair_impl<_Iterators...>::type;
+
+struct __zip_iter_functors
+{
+  // iterator functions
+  _CCCL_EXEC_CHECK_DISABLE
+  template <class _Tuple1, class _Tuple2, size_t... _Indices>
+  _CCCL_API static constexpr bool
+  __iter_op_eq(const _Tuple1& __tuple1, const _Tuple2& __tuple2, _CUDA_VSTD::index_sequence<_Indices...>) noexcept(
+    noexcept(((_CUDA_VSTD::get<_Indices>(__tuple1) == _CUDA_VSTD::get<_Indices>(__tuple2)) || ...)))
+  {
+    return ((_CUDA_VSTD::get<_Indices>(__tuple1) == _CUDA_VSTD::get<_Indices>(__tuple2)) || ...);
+  }
+
+  _CCCL_EXEC_CHECK_DISABLE
+  template <class _Tuple1, class _Tuple2>
+  _CCCL_API static constexpr bool
+  __iter_op_eq(const _Tuple1& __tuple1, const _Tuple2& __tuple2) noexcept(noexcept(__zip_iter_functors::__iter_op_eq(
+    __tuple1,
+    __tuple2,
+    _CUDA_VSTD::make_index_sequence<_CUDA_VSTD::tuple_size_v<_CUDA_VSTD::remove_cvref_t<_Tuple1>>>())))
+  {
+    return __zip_iter_functors::__iter_op_eq(
+      __tuple1,
+      __tuple2,
+      _CUDA_VSTD::make_index_sequence<_CUDA_VSTD::tuple_size_v<_CUDA_VSTD::remove_cvref_t<_Tuple1>>>());
+  }
+
+  struct __zip_op_star
+  {
+    template <class... _Types>
+    using __ret = __tuple_or_pair<decltype(*_CUDA_VSTD::declval<_Types>())...>;
+
+    _CCCL_EXEC_CHECK_DISABLE
+    template <class... _Types>
+    [[nodiscard]] _CCCL_API constexpr __ret<_Types...> operator()(_Types&&... __tuple_elements) const
+      noexcept(noexcept(__ret<_Types...>{*_CUDA_VSTD::forward<_Types>(__tuple_elements)...}))
+    {
+      return __ret<_Types...>{*_CUDA_VSTD::forward<_Types>(__tuple_elements)...};
+    }
+  };
+
+  struct __zip_op_increment
+  {
+    _CCCL_EXEC_CHECK_DISABLE
+    template <class... _Types>
+    _CCCL_API constexpr void operator()(_Types&&... __tuple_elements) const
+      noexcept(noexcept(((void) ++_CUDA_VSTD::forward<_Types>(__tuple_elements), ...)))
+    {
+      ((void) ++_CUDA_VSTD::forward<_Types>(__tuple_elements), ...);
```

Comment:
If anyone is curious, this guards against evil iterators that overload `operator,` 

For example our test iterators ;)